provisioner/data.js

```diff
@@ -1,30 +1,704 @@
+  });
+
+  p = p.then(function() {
+    return names;
+  });
+
+  return p;
+};
+
+
+/**
+ * Turn off every single EC2 instance and cancel all spot
+ * requests which were created by this Provisioner
+ */
+WorkerType.killEverything = function (debug) {
+  assert(debug);
+
+  var p = WorkerType.loadAll.call(this);
+
+  p = p.then(function (workerTypes) {
+    return Promise.all(workerTypes.map(function(workerType) {
+      return workerType.killAll(debug);
+    }));
+  });
+
+  return p;
+};
+
+
+/**
+ * Load a single workerType by name
+ */
+WorkerType.load = function(workerType) {
+  assert(workerType);
+  return base.Entity.load.call(this, {
+    workerType: workerType
+  });
+};
+
+
+/**
+ * Return an Object for JSON encoding which represents
+ * the data associated with this WorkerType.  This is a
+ * method intended for use in displaying the data associated
+ * with a given workerType
+ */
+WorkerType.prototype.json = function() {
+  return lodash.clone(this.__properties);
+};
+
+
+/**
+ * Return the list of regions which can be operated in.
+ * This is a subset of the regions which a given node
+ * is allowed to run in and the list of regions that the
+ * API is configred to run in.
+ */
+WorkerType.prototype.listRegions = function() {
+  var that = this;
+  return Object.keys(this.regions).filter(function(region) {
+    return that.ec2.regions.indexOf(region) !== -1;
+  }); 
+};
+
+
+/**
+ * We use KeyPair names to determine ownership and workerType
+ * in the EC2 world because we can't tag SpotRequests until they've
+ * mutated into Instances.  This sucks and all, but hey, what else
+ * can we do?  This method checks which regions have the required
+ * KeyPair already and creates the KeyPair in regions which do not
+ * already have it
+ */
+WorkerType.prototype.createKeyPair = function() {
+  var that = this;
+
+  var keyName = this.keyPrefix + this.workerType;
+
+  var p = this.ec2.describeKeyPairs.inRegions(this.listRegions(), {
+    Filters: [{
+      Name: 'key-name',
+      Values: [keyName]
+    }] 
+  });
+
+  p = p.then(function(res) {
+    var toCreate = [];
+
+    that.listRegions().forEach(function(region) {
+      var matchingKey = res[region].KeyPairs[0];
+      if (!matchingKey) {
+        toCreate.push(that.ec2.importKeyPair.inRegion(region, {
+          KeyName: keyName,
+          PublicKeyMaterial: that.pubKey,
+        }));
+      } 
+    });
+    return Promise.all(toCreate);
+  });
+
+  return p;
+
+};
+
+
+/**
+ * Delete a KeyPair when it's no longer needed.  This method
+ * does nothing more and you shouldn't run it until you've turned
+ * everything off.
+ */
+WorkerType.prototype.deleteKeyPair = function() {
+  var that = this;
+
+  var keyName = this.keyPrefix + this.workerType;
+
+  var p = this.ec2.describeKeyPairs({
+    Filters: [{
+      Name: 'key-name',
+      Values: [keyName]
+    }] 
+  });
+
+  p = p.then(function(res) {
+    var toDelete = [];
+
+    that.listRegions().forEach(function(region) {
```

Comment:
Please do a `return Promise.all(that.listRegions().map(...`

I don't care if you split it... but using `forEach` and `push` isn't much fun... `filter` + `map` fits great with `Promise.all`.

In this case you might do:
```
var toDelete = that.listRegions().filter(function(region) {
  return res[region].KeyPairs[0];
});
return Promise.all(toDelete.map(function(region) {
  return that.ec2.deleteKeyPair.inRegion(region, {KeyName: keyName});
});
```